src/outpack/copy_files.py

```diff
@@ -0,0 +1,89 @@
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict
+
+from outpack.location_pull import (
+    location_build_pull_plan,
+    location_pull_files,
+)
+from outpack.metadata import PacketFile
+from outpack.root import OutpackRoot
+from outpack.search_options import SearchOptions
+
+
+@dataclass
+class Plan:
+    id: str
+    name: str
+    files: Dict[str, PacketFile]
+
+
+def copy_files(
+    id: str,
+    files: Dict[str, str],
+    dest: Path,
+    options: SearchOptions,
+    root: OutpackRoot,
+) -> Plan:
+    plan = _plan_copy_files(root, id, files)
+
+    try:
+        for here, there in plan.files.items():
+            root.export_file(id, there.path, here, dest)
+
+    except FileNotFoundError as e:
+        if not options.allow_remote:
+            msg = f"File `{e.filename}` from packet {id} is not available locally."
+            raise Exception(msg) from e
+        else:
+            copy_files_from_remote(id, plan.files, dest, options, root)
+
+    return plan
+
+
+def copy_files_from_remote(
+    id: str,
+    files: Dict[str, PacketFile],
+    dest: Path,
+    options: SearchOptions,
+    root: OutpackRoot,
+):
+    plan = location_build_pull_plan(
+        [id],
+        options.location,
+        recursive=False,
+        files={id: [f.hash for f in files.values()]},
+        root=root,
+    )
+
+    with location_pull_files(plan.files, root) as store:
+        for here, there in files.items():
+            store.get(there.hash, dest / here, overwrite=True)
+
+
+def _validate_files(files):
+    if isinstance(files, str):
+        files = {files: files}
+    if isinstance(files, list):
+        files = {x: x for x in files}
+    return files
+
+
+def _plan_copy_files(root, id, files):
+    meta = root.index.metadata(id)
+    files = _validate_files(files)
+    known = {f.path: f for f in meta.files}
+
+    plan = {}
+    for here, there in files.items():
+        # TODO: check absolute paths
+        if here.endswith("/") or there.endswith("/"):
+            msg = "Directories not yet supported for export"
```

Comment:
Directories are unfortunately quite nasty to get right - but they also seem to be popular with users. Definitely best done in another PR though...

Reply:
That code was just moved from `helpers.py`